src/kOS/Module/kOSProcessor.cs

```diff
@@ -298,15 +296,71 @@ private void InitUI()
             var options = (UI_ChooseOption)field.uiControlEditor;
 
-            var bootFiles = new List<string>();
+            var bootFiles = new List<VolumePath>();
 
-            bootFiles.Add("None");
             bootFiles.AddRange(BootDirectoryFiles());
 
             //no need to show the control if there are no available boot files
             options.controlEnabled = bootFiles.Count > 1;
             field.guiActiveEditor = bootFiles.Count > 1;
-            options.options = bootFiles.ToArray();
+            var availableOptions = bootFiles.Select(e => e.ToString()).ToList();
+            var availableDisplays = bootFiles.Select(e => e.Name).ToList();
+
+            availableOptions.Insert(0, "None");
+            availableDisplays.Insert(0, "None");
+
+            var bootFilePath = BootFilePath; // store the selected path temporarily
+            if (bootFilePath != null && !bootFiles.Contains(bootFilePath))
+            {
+                // if the path is not null ("None", null, or empty) and if it isn't in list of files
+                var archive = new Archive(SafeHouse.ArchiveFolder);
+                var file = archive.Open(bootFilePath) as VolumeFile;  // try to open the file as saved
+                if (file == null)
+                {
+                    // check the same file name, but in the boot directory.
+                    var path = VolumePath.FromString(BootDirectoryName).Combine(bootFilePath.Name);
+                    file = archive.Open(path) as VolumeFile; // try to open the new path
+                    if (file == null)
+                    {
+                        // try the file name without "boot" prefix
+                        var name = bootFilePath.Name;
+                        if (name.StartsWith("boot", StringComparison.OrdinalIgnoreCase) && name.Length > 4)
+                        {
+                            // strip the boot prefix and try that file name
+                            name = name.Substring(4);
+                            path = VolumePath.FromString(BootDirectoryName).Combine(name);
+                            file = archive.Open(path) as VolumeFile;  // try to open the new path
+                            if (file == null)
+                            {
+                                // try the file name without "boot_" prefix
+                                if (name.StartsWith("_", StringComparison.OrdinalIgnoreCase) && name.Length > 1)
+                                {
+                                    // only need to strip "_" here
+                                    name = name.Substring(1);
+                                    path = VolumePath.FromString(BootDirectoryName).Combine(name);
+                                    file = archive.Open(path) as VolumeFile;  // try to open the new path
+                                }
+                            }
+                        }
+                    }
+                }
```

Comment:
This looks much easier to follow now.  Thanks.

So, to verify, is this the intended behavior? :

If boot filename can be found, as is, starting from a path that begins in the ``boot/`` directory, use that one first.  (Note, the reason I say  "begins in" that directory rather than saying the file "is in" that directory, is to clarify the meaning if the filename is given as something like ``../myfile.ks``, what the meaning of the ``..`` should end up being.  For the sake of correct documentation, this difference could matter.  Saying it "is not allowed to" have path information in the name would be quite different from saying it "is not necessary to" have path information in the filename.)

Else if boot filename can be found from a path that begins in the archive directory, use that one.  (again, "begins from" having the same meaning as above.)

Else if the filename starts with ``boot_*`` (and thus contains no path information like ``../`` or ``0:/``), then try looking for a file in the ``boot/`` directory that would match the name if the ``boot_`` prefix was stripped off.

Else it has failed to find the file.

Is that the intended behavior?  Because that's what I think it does from the above code path, if I'm understanding it correctly.